### PR TITLE
Charset

### PR DESCRIPTION
**Pull request by mark-bradshaw**
Adding in charset() and charsets(), which are equivalent to encoding() and encodings(), except for the accept-charset header.  My intent is to cover all of the accept headers one by one until they are all parseable by accept.  This is the first.

@hueniverse Now that I'm the lead maintainer, do I need to continue running PRs past you or anyone else, or should I feel free to add additional functionality as I see fit?

### CODE REVIEW

**Comment by mark-bradshaw** [comment]
And just as an FYI, I did a git move on lib/index.js -> lib/encoding.js, but the commit is showing it as if I just changed the file content.  I didn't change anything in encoding.js.  Just moved it.

File: lib/charset.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
+'use strict';
```
Comment (hueniverse) [inline lib/charset.js]: In general, we don't use this in hapi modules. It doesn't really add value.

File: lib/charset.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+exports.charset = function (header, preferences) {
+
+    Hoek.assert(!preferences || Array.isArray(preferences), 'Preferences must be an array');
+    var charsets = header
+        .split(',')
+        .map(internals.getParts)
+        .filter(internals.removeEmpty)
+        .sort(internals.compareByWeight);
+
+    // tack on a default return
```
Comment (hueniverse) [inline lib/charset.js]: Upper case first character on comments and empty line after to match the hapi code style.

File: lib/charset.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    // lower case all preferences
+    preferences = preferences.map(internals.lowerCase);
+
+    // remove any disallowed preferences
+    internals.removeDisallowedPreferences(charsets, preferences);
+
+    // if charsets includes * (that isn't disallowed *;q=0) return first preference
+    var splatLocation = internals.findCharsetItem(charsets, '*');
+    if (~splatLocation && charsets[splatLocation].weight > 0) {
```
Comment (hueniverse) [inline lib/charset.js]: Don't use `~`. It's an ugly hack...

File: lib/charset.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        return preferences[0];
+    }
+
+    // try to find the first match in the array of preferences, ignoring case.
+    for (var i = 0; i < charsets.length; ++i) {
```
Comment (hueniverse) [inline lib/charset.js]: Code style requires assigning length to `il`.

File: lib/charset.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        if (~preferences.indexOf(charsets[i].charset.toLowerCase()) && charsets[i].weight > 0) {
+            return charsets[i].charset;
+        }
+    }
+
+    return '';
+};
+
```
Comment (hueniverse) [inline lib/charset.js]: Two empty lines.

File: lib/charset.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+internals.getParts = function (item) {
+
+    var result = {
+        weight: 1,
+        charset: ''
+    };
+
+    var match = item.match(internals.partsRegex);
+
```
Comment (hueniverse) [inline lib/charset.js]: Reads better without an empty line here.

**Comment by hueniverse** [comment]
@mark-bradshaw in general, it is best to get your code reviewed. Because this is a core dependency, I will be looking at every code change made before updating the dependency in hapi. What I do is merge small changes myself but leave major refactors or new functionality to others to review and merge for me. You need to figure out where's the line for you.

Another thing is that you are new to the hapi coding style and as you can see from my comments (which are just a sample of the issues) there is plenty of getting used to here. It might be helpful to keep doing PR's and having someone who is on the hapi core group review is for you to help get you up to speed.

Completely your call. If you fuck something up here, I'll tell you... :)

**Comment by mark-bradshaw** [comment]
Ok, cool.  Thanks Eran.  I've made adjustments as you've requested.  Anything else you'd like to see?

File: lib/charset.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    header = header.toLowerCase();
+
+    return header
+        .split(',')
+        .map(internals.getParts)
+        .filter(internals.removeEmptyAndDisallowed)
+        .sort(internals.compareByWeight)
+        .map(internals.partToCharset);
```
Comment (hueniverse) [inline lib/charset.js]: Might be worth trying this out with a for loop and a single sort after for performance. My experience is that this much functional JS tends to be very slow. You can look into it later in another PR.

File: lib/charset.js [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        charset: ''
+    };
+
+    var match = item.match(internals.partsRegex);
+    if (!match) {
+        return result;
+    }
+
+    result.charset = match[1];
+    if (match[2] && internals.isNumber(match[2]) ) {
```
Comment (hueniverse) [inline lib/charset.js]: Extra whitespace after `[2])`

File: lib/charset.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+};
+
+
+internals.removeDisallowedPreferences = function (charsets, preferences) {
+
+    var location;
+    for (var i = 0, il = charsets.length; i < il; i++) {
+        if (charsets[i].weight === 0) {
+            location = preferences.indexOf(charsets[i].charset.toLowerCase());
+            if (~location) {
```
Comment (hueniverse) [inline lib/charset.js]: `~` again...

File: test/charset.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
+// Load modules
+
+var Code = require('code');
+var Lab = require('lab');
+var Accept = require('..');
```
Comment (hueniverse) [inline test/charset.js]: Accept comes first (alphabetical)

File: test/charset.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+    it('requires that preferences parameter must be an array', function (done) {
+
+        try {
+            var charset = Accept.charset('iso-8859-5; q=0.1, unicode-1-1;q=0.8, *;q=0.001', 'iso-8859-5');
+        }
+        catch (err) {
+            expect(err).to.exist();
+            return done();
+        }
```
Comment (hueniverse) [inline test/charset.js]:
```js
expect(function () {

    Accept.charset('iso-8859-5; q=0.1, unicode-1-1;q=0.8, *;q=0.001', 'iso-8859-5');
}).to.throw('error_message_here');
```

**Comment by hueniverse** [comment]
A few more comments. Otherwise looks fine.

File: test/charset.js (hunk 11 above)
Comment (mark-bradshaw) [inline test/charset.js]: Thanks!  Much better!

**Comment by mark-bradshaw** [comment]
Per your comments, I've made final cleanup.  Please merge at will.

File: lib/charset.js [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        .map(internals.getParts)
+        .filter(internals.removeEmpty)
+        .sort(internals.compareByWeight);
+
+    // Tack on a default return
+
+    charsets.push({
+            weight: 0.001,
+            charset: ''
+        });
```
Comment (hueniverse) [inline lib/charset.js]: Wrong indentation.

File: lib/charset.js [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+internals.removeEmptyAndDisallowed = function (item) {
+
+    return item.charset !== '' && item.weight !== 0;
+};
+
+
+internals.removeDisallowedPreferences = function (charsets, preferences) {
+
+    var location;
```
Comment (hueniverse) [inline lib/charset.js]: better moved var inside the for loop.

**Comment by hueniverse** [comment]
Merged, but left two more comments.

**Comment by mark-bradshaw** [comment]
Thank you sir.

**Pull request merged**